vital/types/geo_point.h

```diff
@@ -28,105 +28,114 @@
-   *
-   * \returns \c true if object is default constructed.
-   */
-  bool is_empty() const;
-
-protected:
-
-  int m_original_crs;
-  mutable std::unordered_map< int, geo_raw_point_t > m_loc;
-};
-
-VITAL_EXPORT ::std::ostream& operator<< ( ::std::ostream& str, geo_point const& obj );
-
-} } // end namespace
+  namespace vital {
+
+    // ----------------------------------------------------------------------------
+    /** Geo-coordinate.
+     *
+     * This class represents a geolocated point. The point is created by specifying
+     * a raw location and a CRS. The original location and original CRS may be
+     * directly accessed, or the location in a specific CRS may be requested.
+     * Requests for a specific CRS are cached, so that CRS conversion does not need
+     * to be performed every time.
+     *
+     * The CRS values shall correspond to geodetic CRS's as specified by the
+     * European Petroleum Survey Group (EPSG) Spatial Reference System Identifiers
+     * (SRID's). Some well known values are defined by kwiver::vital::SRID.
+     *
+     * Note that the underlying values are ordered easting, northing, for
+     * consistency with Euclidean convention (X, Y), and \em not northing, easting
+     * as is sometimes used for geo-coordinates.
+     *
+     * \see https://en.wikipedia.org/wiki/Spatial_reference_system,
+     *      http://www.epsg.org/, https://epsg-registry.org/
+     */
+    class VITAL_EXPORT geo_point
+    {
+    public:
+      using geo_raw_point_t = kwiver::vital::vector_3d;
+
+      geo_point();
+      geo_point(geo_raw_point_t const&, int crs);
+
+      virtual ~geo_point() = default;
+
+      /**
+       * \brief Accessor for location in original CRS.
+       *
+       * \returns The location in the CRS that was used to set the location.
+       * \throws std::out_of_range Thrown if no location has been set.
+       *
+       * \see crs()
+       */
+      geo_raw_point_t location() const;
+
+      /**
+       * \brief Accessor for original CRS.
+       *
+       * \returns The CRS used to set the location.
+       *
+       * \see location()
+       */
+      int crs() const;
+
+      /**
+       * \brief Accessor for the location.
+       *
+       * \returns The location in the requested CRS.
+       * \throws std::runtime_error if the conversion fails.
+       */
+      geo_raw_point_t location(int crs) const;
+
+      /**
+       * \brief Set location.
+       *
+       * This sets the geo-coordinate to the specified location, which is defined
+       * by the raw location and specified CRS.
+       */
+      void set_location(geo_raw_point_t const&, int crs);
+
+      /**
+       * \brief Test if point has a specified location.
+       *
+       * This method checks the object to see if any location data has been set.
+       *
+       * \returns \c true if object is default constructed.
+       */
+      bool is_empty() const;
+
+      covariance_3f& covariance() { return m_covariance; }
```

Comment:
This methods really need documentation.  How is covariance defined for a geo_point?  The values are often not in a cartesian coordinate system.  I'm not convinced that covariance belongs on this class, and if it does, it is critical that we have documentation to describe how to interpret those values.

Reply:
Given that there are many use cases for a geo_point without covariance, I'd much rather seen separate class for representing a geo_point plus covariance for the cases that need that.

Reply:
@mattbrown11  Can you help out with this?

Reply:
can we just make the covariance a shared_ptr so that its optional?

Reply:
No.  That is not how shared_ptr should be used for small objects like this.  This is not as much a memory constraint as it a design issue.  Covariance might be something use need in tracking, but it's not what defines a geo_point.   It seems like you would want an "uncertain_geo_point" or "geo_point_estimate" or something like that which contains a geo_point and covariance.

Reply:
Ok, current suggested name from Matt B is 'geo_point_with_cov'

Currently, taking all applications